cli/download_source.go

```diff
@@ -0,0 +1,259 @@
+}
+
+// 1. Download the given source URL, which should use Terraform's module source syntax, into a temporary folder
+// 2. Copy the contents of terragruntOptions.WorkingDir into the temporary folder.
+// 3. Set terragruntOptions.WorkingDir to the temporary folder.
+//
+// See the processTerraformSource method for how we determine the temporary folder so we can reuse it across multiple
+// runs of Terragrunt to avoid downloading everything from scratch every time.
+func downloadTerraformSource(source string, terragruntOptions *options.TerragruntOptions) error {
+	terraformSource, err := processTerraformSource(source, terragruntOptions)
+	if err != nil {
+		return err
+	}
+
+	if err := downloadTerraformSourceIfNecessary(terraformSource, terragruntOptions); err != nil {
+		return err
+	}
+	
+	terragruntOptions.Logger.Printf("Copying files from %s into %s", terragruntOptions.WorkingDir, terraformSource.DownloadDir)
+	if err := util.CopyFolderContents(terragruntOptions.WorkingDir, terraformSource.DownloadDir); err != nil {
+		return err
+	}
+
+	terragruntOptions.Logger.Printf("Setting working directory to %s", terraformSource.DownloadDir)
+	terragruntOptions.WorkingDir = terraformSource.DownloadDir
+
+	return nil
+}
+
+// Download the specified TerraformSource if the latest code hasn't already been downloaded.
+func downloadTerraformSourceIfNecessary(terraformSource *TerraformSource, terragruntOptions *options.TerragruntOptions) error {
+	alreadyLatest, err := alreadyHaveLatestCode(terraformSource)
+	if err != nil {
+		return err
+	}
+
+	if alreadyLatest {
+		terragruntOptions.Logger.Printf("Terraform files in %s are up to date. Will not download again.", terraformSource.DownloadDir)
+		return nil
+	}
+
+	if err := cleanupTerraformFiles(terraformSource.DownloadDir, terragruntOptions); err != nil {
+		return err
+	}
+
+	if err := terraformInit(terraformSource, terragruntOptions); err != nil {
+		return err
+	}
+
+	if err := writeVersionFile(terraformSource); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// Returns true if the specified TerraformSource, of the exact same version, has already been downloaded into the
+// DownloadFolder. This helps avoid downloading the same code multiple times. Note that if the TerraformSource points
+// to a local file path, we assume the user is doing local development and always return false to ensure the latest
+// code is downloaded (or rather, copied) every single time. See the processTerraformSource method for more info.
+func alreadyHaveLatestCode(terraformSource *TerraformSource) (bool, error) {
+	if 	isLocalSource(terraformSource.CanonicalSourceURL) ||
+		!util.FileExists(terraformSource.DownloadDir) ||
+		!util.FileExists(terraformSource.VersionFile) {
+
+		return false, nil
+	}
+
+	currentVersion := encodeSourceVersion(terraformSource.CanonicalSourceURL)
```

Comment:
Yes, I considered this. In the future, if this is a common case, we could add a `terragrunt-update` or similar flag. For now, the trivial workaround is to delete the relevant tmp folder manually. My hope, however, is that this is a relatively rare case.